CASMCMS-8231: BOS - `rootfs_provider` parameter explanation (#2491)
Cray-HPE/docs-csm: operations/boot_orchestration/Session_Templates.md

    Consult the `cray-hms-base-config` Kubernetes ConfigMap in the `services` namespace for a listing of the available roles and sub-roles on the system.

#### Rootfs Providers

>   The 'rootfs' is the root file system.

Reviewer: ```suggestion
The `rootfs` is the root file system.
```

>   `rootfs_provider` identifies the mechanism that providers the root file system for the node. 

Reviewer 2: "provides" vs "providers"

>   In the case of the Cray Operating System (COS) image, the rootfs provider is HPE’s Content Projection Service (CPS), which uses HPE’s Data Virtualization Service (DVS) to deliver the content. CPS projects the root file system onto the nodes as a squashfs image. This is provided via an overlay file system which is set up in dracut. 

Reviewer: ```suggestion
In the case of the Cray Operating System (COS) image, the `rootfs` provider is HPE’s Content Projection Service (CPS), which uses HPE’s Data Virtualization Service (DVS) to deliver the content. CPS projects the root file system onto the nodes as a SquashFS image. This is provided via an overlay file system which is set up in dracut. 
```

Author: I addressed this.

>   In the case of the Cray Operating System (COS) image, the rootfs provider is HPE’s Content Projection Service (CPS), which uses HPE’s Data Virtualization Service (DVS) to deliver the content. CPS projects the root file system onto the nodes as a squashfs image. This is provided via an overlay file system which is set up in dracut. 

Reviewer 3: ```suggestion
In the case of the Cray Operating System (COS) image, the rootfs provider is HPE’s Content Projection Service (CPS), which uses HPE’s Data Virtualization Service (DVS) to deliver the content. CPS projects the root file system onto the nodes as a squashfs image. This is provided via an overlay file system which is set up in dracut. 
```

>   `rootfs_provider_passthrough` is a string that is passed through to the provider of the rootfs. This string can contain additional information that the provider will act upon. 

Reviewer: ```suggestion
`rootfs_provider_passthrough` is a string that is passed through to the provider of the `rootfs`. This string can contain additional information that the provider will act upon. 
```

>   Both the `rootfs_provider` and `rootfs_provider_passthrough` parameters are used to construct the value of the kernel boot parameter `root` that BOS sends to the node.

Reviewer: ```suggestion
Both the `rootfs_provider` and `rootfs_provider_passthrough` parameters are used to construct the value of the kernel boot parameter `root` that BOS sends to the node.
```

>   BOS constructs the kernel boot parameter `root` per the following syntax.

Reviewer: ```suggestion
BOS constructs the kernel boot parameter `root` per the following syntax.
```

>   ```bash

Reviewer: ````suggestion
```text
````

>   root=<Protocol>:<Root FS location>:<Etag>:<RootFS-provider-passthrough parameters>

Reviewer: ```suggestion
root=<Protocol>:<Root FS location>:<Etag>:<RootFS-provider-passthrough parameters>
```

>   ```

Reviewer: ````suggestion
```
````

>   BOS fills in the protocol based off of the value provided in `rootfs_provider`. If BOS does not know the rootfs_provider, it omits the Protocol field. Currently, BOS only recognizes the rootfs_provider ‘cpss3’.

Reviewer: ```suggestion
BOS fills in the protocol based on the value provided in `rootfs_provider`. If BOS does not know the `rootfs_provider`, then it omits the protocol field. Currently, BOS only recognizes the `rootfs_provider` `cpss3`.
```

>   BOS finds the ‘Root FS provider’ and ‘Etag’ values in the manifest file in the session template in the [boot set](#boot-artifacts-and-s3).

Reviewer: ```suggestion
BOS finds the `Root FS provider` and `Etag` values in the manifest file in the session template in the [boot set](#boot-artifacts-and-s3).
```

>   The rootfs_provider_passthrough parameters are appended to the `root` parameter without modification. They are ‘passed through’ as the name imples.

Reviewer: ```suggestion
The `rootfs_provider_passthrough` parameters are appended to the `root` parameter without modification. They are "passed through", as the name imples.
```

>   `root` kernel parameter example:

Reviewer: ```suggestion
##### `root` kernel parameter example
```

>   ```bash
>   root=craycps-s3:s3://boot-images/b9caaf66-c0b4-4231-aba7-a45f6282b21d/rootfs:f040d70bd6fabaf91838fe4e484563cf-211:dvs:api-gw-service-nmn.local:300:nmn0 nmd_data=url=s3://boot-images/b9caaf66-c0b4-4231-aba7-a45f6282b21d/rootfs,etag=f040d70bd6fabaf91838fe4e484563cf-211
>   ```
> 
>   The following table dissects and explains the different pieces in the preceding example.

Reviewer: ````suggestion
```text
root=craycps-s3:s3://boot-images/b9caaf66-c0b4-4231-aba7-a45f6282b21d/rootfs:f040d70bd6fabaf91838fe4e484563cf-211:dvs:api-gw-service-nmn.local:300:nmn0 nmd_data=url=s3://boot-images/b9caaf66-c0b4-4231-aba7-a45f6282b21d/rootfs,etag=f040d70bd6fabaf91838fe4e484563cf-211
```

The following table explains the different pieces in the preceding example.
````

>   **Table:** kernel boot parameter `root` syntax

Reviewer: ```suggestion
```

  |Field|Example Value|Explanation|
  |-----|-------------|-----------|

>   |Protocol|craycps-s3|The protocol used to mount the root file system, using CPS|

Reviewer 4: I wasn't clear when I first read this if this explanation field always refers to a CPS protocol or not.  Is that true, or is this only using CPS because it's in the protocol name?

Author: I attempted to make this clearer. Let me know if I succeeded.

>   |Field|Example Value|Explanation|
>   |-----|-------------|-----------|
>   |Protocol|craycps-s3|The protocol used to mount the root file system, using CPS|
>   |RootFS Provider Location|s3://boot-images/b9caaf66-c0b4-4231-aba7-a45f6282b21d/rootfs|The rootfs provider location is a squashfs image stored in S3|
>   |Etag|f040d70bd6fabaf91838fe4e484563cf-211|The Etag (entity tag) is the identifier of the squashfs image in S3.|
>   |Rootfs provider passthrough parameters|dvs:api-gw-service-nmn.local:300:nmn0|These are additional parameters that CPS uses to properly mount the file system|

Reviewer: ```suggestion
|Field|Example Value|Explanation|
|-----|-------------|-----------|
|Protocol|`craycps-s3`|The protocol used to mount the root file system, using CPS|
|`rootfs` Provider Location|`s3://boot-images/b9caaf66-c0b4-4231-aba7-a45f6282b21d/rootfs`|The `rootfs` provider location is a SquashFS image stored in S3|
|`Etag`|`f040d70bd6fabaf91838fe4e484563cf-211`|The `Etag` (entity tag) is the identifier of the SquashFS image in S3.|
|`rootfs` provider passthrough parameters|`dvs:api-gw-service-nmn.local:300:nmn0`|These are additional parameters that CPS uses to properly mount the file system|
```

>   The rootfs_provider_passthrough parameters are explained in the following table.
> 
>   **Table:** Rootfs provider passthrough parameters
> 
>   |Parameter|Example|Explanation|
>   |---|---|---|
>   |Transport|dvs|Use DVS to project the squashfs image down to the node|
>   |Gateway|api-gw-service-nmn.local|This is the URL that identifies the gateway where the DVS servers are located|
>   |Time-out|300|The number of seconds to wait to establish a contact|
>   |Interface|nmn0|The IP interface on the node to use to contact the DVS server; This interface must be up to continue booting.|

Reviewer: ```suggestion
The `rootfs_provider_passthrough` parameters are explained in the following table.

|Parameter|Example|Explanation|
|---|---|---|
|Transport|`dvs`|Use DVS to project the SquashFS image down to the node|
|Gateway|`api-gw-service-nmn.local`|This is the URL that identifies the gateway where the DVS servers are located|
|Time-out|`300`|The number of seconds to wait to establish a contact|
|Interface|`nmn0`|The IP interface on the node to use to contact the DVS server; This interface must be up to continue booting.|
```

  Note:

>   Regarding the interface to use for contacting DVS, the possible values are

Reviewer 3: ```suggestion
  Regarding the interface to use for contacting DVS, the possible values are

```

>   Note:
>   Regarding the interface to use for contacting DVS, the possible values are
>   * `nmn0` -- Ensures the nmn0 interface is up
>   * `nmn0,hsn0` -- Ensures both the nmn0 and hsn0 interfaces are up. This is required for booting over the HSN.
>   * `hsn0` -- Ensures the hsn0 interface is up. 
>   
>   The DVS configuration files determine what interface to use (NMN or HSN). However, the CPS dracut ensures the requested interfaces are up.

Reviewer: ```suggestion
Note:
Regarding the interface to use for contacting DVS, the possible values are:

  * `nmn0` -- Ensures that the `nmn0` interface is up
  * `nmn0,hsn0` -- Ensures that both the `nmn0` and `hsn0` interfaces are up. This is required for booting over the HSN.
  * `hsn0` -- Ensures that the `hsn0` interface is up. 
  
The DVS configuration files determine what interface to use (NMN or HSN). However, the CPS dracut ensures that the requested interfaces are up.
```

Author: Fixed.
